anomalib/models/fastflow/torch_model.py

```diff
@@ -0,0 +1,250 @@
+        return loss
+
+
+class FastflowModel(nn.Module):
+    """FastFlow.
+
+    Unsupervised Anomaly Detection and Localization via 2D Normalizing Flows.
+
+    Args:
+        input_size (Tuple[int, int]): Model input size.
+        backbone (str): Backbone CNN network
+        flow_steps (int): Flow steps.
+        conv3x3_only (bool, optinoal): Use only conv3x3 in fast_flow model. Defaults to False.
+        hidden_ratio (float, optional): Ratio to calculate hidden var channels. Defaults to 1.0.
+
+    Raises:
+        ValueError: When the backbone is not supported.
+    """
+
+    def __init__(
+        self,
+        input_size: Tuple[int, int],
+        backbone: str,
+        flow_steps: int,
+        conv3x3_only: bool = False,
+        hidden_ratio: float = 1.0,
+    ) -> None:
+        super().__init__()
+
+        self.input_size = input_size
+
+        if backbone in ["cait_m48_448", "deit_base_distilled_patch16_384"]:
+            self.feature_extractor = timm.create_model(backbone, pretrained=True)
+            channels = [768]
+            scales = [16]
+        elif backbone in ["resnet18", "wide_resnet50_2"]:
+            self.feature_extractor = timm.create_model(
+                backbone,
+                pretrained=True,
+                features_only=True,
+                out_indices=[1, 2, 3],
+            )
+            channels = self.feature_extractor.feature_info.channels()
+            scales = self.feature_extractor.feature_info.reduction()
+
+            # for transformers, use their pretrained norm w/o grad
+            # for resnets, self.norms are trainable LayerNorm
+            self.norms = nn.ModuleList()
+            for channel, scale in zip(channels, scales):
+                self.norms.append(
+                    nn.LayerNorm(
+                        [channel, int(input_size[0] / scale), int(input_size[1] / scale)],
+                        elementwise_affine=True,
+                    )
+                )
+        else:
+            raise ValueError(
+                f"Backbone {backbone} is not supported. List of available backbones are "
+                "[cait_m48_448, deit_base_distilled_patch16_384, resnet18, wide_resnet50_2]."
+            )
+
+        for parameter in self.feature_extractor.parameters():
+            parameter.requires_grad = False
+
+        self.fast_flow_blocks = nn.ModuleList()
+        for channel, scale in zip(channels, scales):
+            self.fast_flow_blocks.append(
+                create_fast_flow_block(
+                    input_dimensions=[channel, int(input_size[0] / scale), int(input_size[1] / scale)],
+                    conv3x3_only=conv3x3_only,
+                    hidden_ratio=hidden_ratio,
+                    flow_steps=flow_steps,
+                )
+            )
+        self.anomaly_map_generator = AnomalyMapGenerator(input_size=input_size)
+
+    def forward(self, x: Tensor) -> Union[Tuple[List[Tensor], List[Tensor]], Tensor]:
```

Comment:
This method is a bit hard to read. Maybe you could split it into several sub-methods. Something like:
```
if isinstance(self.feature_extractor, VisionTransformer):
    features = self.get_transformer_features(x)
elif isinstance(self.feature_extractor, Cait):
    features = self.get_cait_features(x)
else:
    features = self.feature_extractor(x)
    
hidden_variables = self.get_hidden_variables(features)
```